TruffleString-Experiment.md

```diff
@@ -1,3 +1,284 @@
+
+The goal of the second experiment was to reduce the code size of the `String>>#compareWith:` method. The code size of the `String>>#compareWith:` method for the second experiment is as follows:
+
+| Metric    | Tier | Baseline    | Experiment 1 | Experiment 2  | Change (%) |
+| --------- | ---- | ----------- | ------------ | ------------- | ---------- |
+| Code Size | 1    | 2 040 bytes | 10 840 bytes | 8 276 bytes ⬆️ | -23.6% ⬇️   |
+| Code Size | 2    | 897 bytes   | 6 839 bytes  | 4 403 bytes   | -36.0% ⬇️   |
+
+We can see that the code size of the `String>>#compareWith:` method decreased significantly in the second experiment compared to the first experiment. This indicates that the method was optimized better, leading to a smaller code size and improved performance.
+
+The compilation graph of `String>>#compareWith:` supports that. The number of compilation steps decreased. The fist image contains the compilation graph for the first experiment, while the second image contains the compilation graph for the second experiment:
+
+![String>>#compareWith: Compilation Graph - Experiment 1](img/igv-experiment-1.png)
+
+![String>>#compareWith: Compilation Graph - Experiment 2](img/igv-experiment-2.png)
+
+The visual comparison confirms that eliminating `ByteArray` and `ByteSymbol` type permutations simplified the compilation graph, reducing the number of specialized code paths the compiler must generate.
+
+Despite the compilation improvements, the JSON benchmark performance remains 45.4% slower than baseline. This persistent performance gap stems from two fundamental limitations:
+
+**1. TruffleString Optimization Mismatch:**
+`TruffleStrings` are optimized for large strings through vector-based representation and parallelism (> 50 characters). However, Squeak frequently processes short strings like method names (represented as `ByteSymbols`), where TruffleString's sophisticated optimizations create overhead rather than benefits [4].
+
+**2. Primitive Operation Granularity:**
+Smalltalk primitives are atomic and focus on byte-level operations. The most complex string operations are handled by the Squeak VM itself. This architectural constraint means we primarily use basic read/write operations on `TruffleString`, which cannot leverage its advanced features like vectorization and parallelism.
+
+To validate that TruffleString can provide significant performance benefits when used appropriately, we conducted a targeted micro-benchmark focusing on large string comparisons—TruffleString's optimal use case.
+
+This benchmark compares two identical 10-million character strings using TruffleString's native `CompareBytes` function instead of Squeak's byte-level implementation:
+
+```smalltalk
+|a b|
+a := String new: 10000000 withAll: $a.
+b := a copy.
+[100 timesRepeat: [a = b]] timeToRunWithoutGC
+```
+
+We ran this micro benchmark on the same machine as the JSON benchmark. The results are as follows:
```

Comment:
Should we also denote the OSVM performance here? I remember that we achieved better results than the OSVM now 😉 🥳

Reply:
Great idea! I will add it and the used versions of osvm and tsvm 👍